Android back button override closes app

The Android back button override from PrismLibrary/Prism.Maui#88 isn't working properly on my end.

### Steps to Reproduce

1. Build and launch _PrismMauiDemo_ on an Android 10.0 emulator.
2. Click _Flyout Page_.
3. Click the _Back_ button on the emulator.

### Expected Behavior

- Navigation from _ViewD_ to _ViewC_:

```
[DOTNET] ViewD - View Disappearing
[DOTNET] ViewC - View Appearing
[DOTNET] ViewD - ViewModel NavigatedFrom
[DOTNET] ViewC - ViewModel NavigatedTo
[DOTNET] Navigation: GoBack
[DOTNET] Result: Success
```

### Observed Behavior

- The app closes, and all manner of badness ensues:

```
[DOTNET] ViewD - View Disappearing
[DOTNET] ViewC - View Appearing
[DOTNET] ViewC - View Disappearing
[DOTNET] ViewD - ViewModel NavigatedFrom
[DOTNET] ViewC - ViewModel NavigatedTo
[DOTNET] Navigation: GoBack
[DOTNET] Result: Success
[DOTNET] ViewC - ViewModel NavigatedFrom
[DOTNET] ViewB - ViewModel NavigatedTo
[DOTNET] Navigation: GoBack
[DOTNET] Result: Success
```

- Reopening the app from recent apps reveals _ViewB_ with no navigation bar:

`[DOTNET] ViewB - View Appearing`

- Reopening the app from the launcher throws an exception and crashes:

```
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.NotImplementedException: Either set MainPage or override CreateWindow.
[mono-rt]    at Microsoft.Maui.Controls.Application.CreateWindow(IActivationState activationState) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Application\Application.Impl.cs:line 126
[mono-rt]    at Microsoft.Maui.Controls.Application.Microsoft.Maui.IApplication.CreateWindow(IActivationState activationState) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Application\Application.Impl.cs:line 40
[mono-rt]    at Microsoft.Maui.Platform.ApplicationExtensions.CreatePlatformWindow(Activity activity, IApplication application, Bundle savedInstanceState) in D:\a\_work\1\s\src\Core\src\Platform\Android\ApplicationExtensions.cs:line 44
[mono-rt]    at Microsoft.Maui.MauiAppCompatActivity.OnCreate(Bundle savedInstanceState) in D:\a\_work\1\s\src\Core\src\Platform\Android\MauiAppCompatActivity.cs:line 32
[mono-rt]    at PrismMauiDemo.MainActivity.OnCreate(Bundle savedInstanceState) in C:\Users\NeilThiessen\Downloads\Prism.Maui-master\sample\PrismMauiDemo\Platforms\Android\MainActivity.cs:line 12
[mono-rt]    at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-31/mcw/Android.App.Activity.cs:line 2781
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 121
```

--- a/e2e/Maui/PrismMauiDemo.sln
+++ b/e2e/Maui/PrismMauiDemo.sln
@@ -29,7 +29,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Prism.Maui.Tests", "..\..\t
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Prism.Core.Tests", "..\..\tests\Prism.Core.Tests\Prism.Core.Tests.csproj", "{E0F13AA9-8083-47CA-B10D-93C5285D1505}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Prism.Events", "..\..\src\Prism.Events\Prism.Events.csproj", "{5623CB62-59C1-49BC-BB16-4C5D63D82DAC}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Prism.Events", "..\..\src\Prism.Events\Prism.Events.csproj", "{5623CB62-59C1-49BC-BB16-4C5D63D82DAC}"
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
@@ -95,6 +95,7 @@ Global
 		{EE6F0C99-61D1-4E2E-8185-FBA0D246D5C7} = {E91F80AA-3D61-4C28-B876-3EDFB5921E7D}
 		{F3D2DFDB-95FB-4CBB-A624-35EB6550854D} = {E91F80AA-3D61-4C28-B876-3EDFB5921E7D}
 		{E0F13AA9-8083-47CA-B10D-93C5285D1505} = {E91F80AA-3D61-4C28-B876-3EDFB5921E7D}
+		{5623CB62-59C1-49BC-BB16-4C5D63D82DAC} = {8202B92A-A573-4365-8A15-E246504A7CBD}
 	EndGlobalSection
 	GlobalSection(ExtensibilityGlobals) = postSolution
 		SolutionGuid = {50B0D1F3-D832-4C6C-858E-24F5F3B33632}

--- a/src/Maui/Prism.Maui/Common/MvvmHelpers.cs
+++ b/src/Maui/Prism.Maui/Common/MvvmHelpers.cs
@@ -269,13 +269,42 @@ public static class MvvmHelpers
         if (lastModal != null)
             page = lastModal;
 
-        return GetOnNavigatedToTargetFromChild(page);
+        return EvaluateCurrentPage(page);
     };
+
+    private static Page EvaluateCurrentPage(Page target)
+    {
+        Page child = null;
+
+        if (target is FlyoutPage flyout)
+            child = flyout.Detail;
+        else if (target is TabbedPage tabbed)
+            child = tabbed.CurrentPage;
+        else if (target is NavigationPage np)
+            child = np.Navigation.NavigationStack.Last();
+
+        if (child != null)
+            target = GetOnNavigatedToTargetFromChild(child);
+
+        if (target is Page page)
+            return page.Parent switch
+            {
+                TabbedPage tab when tab.CurrentPage != target => EvaluateCurrentPage(tab.CurrentPage),
+                NavigationPage nav when nav.CurrentPage != target => EvaluateCurrentPage(nav.CurrentPage),
+                _ => target
+            };
+
+        return null;
+    }
 
     public static async Task HandleNavigationPageGoBack(NavigationPage navigationPage)
     {
         var navigationService = Navigation.Xaml.Navigation.GetNavigationService(navigationPage.CurrentPage);
-        await navigationService.GoBackAsync();
+        var result = await navigationService.GoBackAsync();
+        if(result.Exception is NavigationException navEx && navEx.Message == NavigationException.CannotPopApplicationMainPage)
+        {
+            Application.Current.Quit();
+        }
     }
 
     public static void HandleSystemGoBack(IView previousPage, IView currentPage)

--- a/src/Maui/Prism.Maui/Controls/PrismNavigationPage.cs
+++ b/src/Maui/Prism.Maui/Controls/PrismNavigationPage.cs
@@ -2,13 +2,23 @@
 
 namespace Prism.Controls;
 
+/// <summary>
+/// Provides a wrapper for the NavigationPage to better handle the OnBackButtonPressed event with Prism Navigation
+/// </summary>
 public class PrismNavigationPage : NavigationPage
 {
+    /// <summary>
+    /// Creates a new instance of the <see cref="PrismNavigationPage"/>
+    /// </summary>
     public PrismNavigationPage()
     {
         BackButtonPressed += HandleBackButtonPressed;
     }
 
+    /// <summary>
+    /// Creates a new instance of the <see cref="PrismNavigationPage"/> with a specified <see cref="Page"/> at the Root
+    /// </summary>
+    /// <param name="page"></param>
     public PrismNavigationPage(Page page)
         : base(page)
     {
@@ -17,7 +27,6 @@ public class PrismNavigationPage : NavigationPage
 
     /// <inheritdoc/>
     public event EventHandler BackButtonPressed;
-
 
     /// <inheritdoc/>
     protected override bool OnBackButtonPressed()

--- a/src/Maui/Prism.Maui/Controls/PrismNavigationPage.cs
+++ b/src/Maui/Prism.Maui/Controls/PrismNavigationPage.cs
@@ -15,8 +15,11 @@ public class PrismNavigationPage : NavigationPage
         BackButtonPressed += HandleBackButtonPressed;
     }
 
+    /// <inheritdoc/>
     public event EventHandler BackButtonPressed;
 
+
+    /// <inheritdoc/>
     protected override bool OnBackButtonPressed()
     {
         BackButtonPressed.Invoke(this, EventArgs.Empty);

--- a/src/Maui/Prism.Maui/Navigation/PrismWindow.cs
+++ b/src/Maui/Prism.Maui/Navigation/PrismWindow.cs
@@ -1,5 +1,11 @@
-﻿using Prism.AppModel;
+﻿using System.ComponentModel;
+using Prism.AppModel;
 using Prism.Common;
+using Prism.Ioc;
+using Prism.Navigation.Xaml;
+using Prism.Services;
+using Prism.Xaml;
+using TabbedPage = Microsoft.Maui.Controls.TabbedPage;
 
 namespace Prism.Navigation;
 
@@ -18,6 +24,56 @@ internal class PrismWindow : Window
     public bool IsActive { get; private set; }
 
     internal Page CurrentPage => Page is null ? null : MvvmHelpers.GetCurrentPage(Page);
+
+    internal bool IsRootPage => Page switch
+    {
+        TabbedPage tabbed => tabbed.CurrentPage,
+        NavigationPage nav => nav.RootPage,
+        _ => Page
+    } == CurrentPage;
+
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public void OnSystemBack()
+    {
+        var currentPage = CurrentPage;
+        if(currentPage?.Parent is NavigationPage navPage)
+        {
+            // The NavigationPage has already taken care of the GoBack
+            return;
+        }
+
+        var container = currentPage.GetContainerProvider();
+
+        if (IsRoot(currentPage))
+        {
+            var app = container.Resolve<IApplication>() as Application;
+            app.Quit();
+            return;
+        }
+        else if (currentPage is IDialogContainer dialogContainer)
+        {
+            if (dialogContainer.Dismiss.CanExecute(null))
+                dialogContainer.Dismiss.Execute(null);
+        }
+        else
+        {
+            var navigation = container.Resolve<INavigationService>();
+            navigation.GoBackAsync();
+        }
+    }
+
+    private bool IsRoot(Page page)
+    {
+        if (page == Page) return true;
+
+        return page.Parent switch
+        {
+            FlyoutPage flyout => IsRoot(flyout),
+            TabbedPage tabbed => IsRoot(tabbed),
+            NavigationPage navigation => IsRoot(navigation),
+            _ => false
+        };
+    }
 
     private async void PrismWindow_ModalPopping(object sender, ModalPoppingEventArgs e)
     {

--- a/src/Maui/Prism.Maui/PrismAppBuilder.cs
+++ b/src/Maui/Prism.Maui/PrismAppBuilder.cs
@@ -45,28 +45,23 @@ public sealed class PrismAppBuilder
                 {
                     var root = ContainerLocator.Container;
                     if (root is null)
-                        return true;
+                        return false;
 
                     var app = root.Resolve<IApplication>();
                     var windows = app.Windows.OfType<PrismWindow>();
                     if (!windows.Any(x => x.IsActive))
-                        return true;
+                        return false;
 
                     var window = windows.First(x => x.IsActive);
-                    var currentPage = window.CurrentPage;
-                    var container = currentPage.GetContainerProvider();
-                    if(currentPage is IDialogContainer dialogContainer)
+                    if(window.IsRootPage && app is Application application)
                     {
-                        if (dialogContainer.Dismiss.CanExecute(null))
-                            dialogContainer.Dismiss.Execute(null);
-                    }
-                    else
-                    {
-                        var navigation = container.Resolve<INavigationService>();
-                        navigation.GoBackAsync();
+                        application.Quit();
+                        return false;
                     }
 
-                    return false;
+                    window.OnSystemBack();
+
+                    return true;
                 });
             });
 #endif

